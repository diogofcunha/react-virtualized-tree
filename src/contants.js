export const UPDATE_TYPE = {
  ADD: 0,
  DELETE: 1,
  UPDATE: 2,
  MOVE_AS_CHILDREN: 3,
  MOVE_AS_SIBLING: 4
}