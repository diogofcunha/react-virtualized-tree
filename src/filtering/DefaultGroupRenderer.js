import React from 'react';

const DefaultGroupRenderer = ({onChange, groups, selectedGroup}) => {
  return (
    <select
      className="tree-group"
      onChange={({target: {value}}) => {
        onChange(value);
      }}
      value={selectedGroup}
    >
      {Object.keys(groups).map(g => (
        <option key={g} value={g}>
          {groups[g].name}
        </option>
      ))}
    </select>
  );
};

export default DefaultGroupRenderer;
