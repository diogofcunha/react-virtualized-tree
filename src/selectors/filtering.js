const INITIAL_FILTERED_VALUE = {nodes: [], nodeParentMappings: {}};

export const filterNodes = (filter, nodes, parents = []) =>
  nodes.reduce((filtered, n) => {
    const {nodes: filteredChildren, nodeParentMappings: childrenNodeMappings} = n.children
      ? filterNodes(filter, n.children, [...parents, n.id])
      : INITIAL_FILTERED_VALUE;

    return !(filter(n) || filteredChildren.length)
      ? filtered
      : {
          nodes: [
            ...filtered.nodes,
            {
              ...n,
              children: filteredChildren,
            },
          ],
          nodeParentMappings: {
            ...filtered.nodeParentMappings,
            ...childrenNodeMappings,
            [n.id]: parents,
          },
        };
  }, INITIAL_FILTERED_VALUE);
