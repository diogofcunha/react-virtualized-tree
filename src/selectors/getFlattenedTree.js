export const isNodeExpanded = node => node.state && node.state.expanded;
export const nodeHasChildren = node => node.children && node.children.length;

export const getFlattenedTree = (nodes, parents = []) =>
  nodes.reduce((flattenedTree, node) => {
    const deepness = parents.length;
    const nodeWithHelpers = {...node, deepness, parents};

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      return [...flattenedTree, nodeWithHelpers];
    }

    return [...flattenedTree, nodeWithHelpers, ...getFlattenedTree(node.children, [...parents, node.id])];
  }, []);

export const getFlattenedTreePaths = (nodes, parents = []) => {
  const paths = [];

  for (const node of nodes) {
    const {id} = node;

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      paths.push(parents.concat(id));
    } else {
      paths.push(parents.concat(id));
      paths.push(...getFlattenedTreePaths(node.children, [...parents, id]));
    }
  }

  return paths;
};

export const doesChangeAffectFlattenedTree = (previousNode, nextNode) => {
  return isNodeExpanded(previousNode) !== isNodeExpanded(nextNode);
};
