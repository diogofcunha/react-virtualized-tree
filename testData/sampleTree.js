export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
