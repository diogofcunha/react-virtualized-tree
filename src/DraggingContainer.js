import React from 'react';
import PropTypes from 'prop-types';

import { Node } from './shapes/nodeShapes';
import { UPDATE_TYPE } from './contants';

export default class DraggingContainer extends React.Component {
  state = {
    fromNode: {}
  }

  handleDrag = fromNode => {
    this.setState({ fromNode });
  }

  handleDrop = (toNode, type) => {
    this.props.onChange({
      type,
      node: [ this.state.fromNode, toNode ]
    });

    this.setState({ fromNode: {} });
  }

  getChildContext = () => ({
    handleDrag: this.handleDrag,
    handleDrop: this.handleDrop,
    fromNode: this.state.fromNode
  })

  static childContextTypes = {
    handleDrag: PropTypes.func,
    handleDrop: PropTypes.func,
    fromNode: PropTypes.object
  }

  render() {
    return (
      <div
        className={this.props.className}
        onDrop={() => this.handleDrop({ ...this.state.fromNode, parents: [] }, UPDATE_TYPE.MOVE_AS_SIBLING)}
        onDragOver={e => e.preventDefault()}
      > 
        { this.props.children }
      </div>
    )
  }
}

DraggingContainer.propTypes = {
  onChange: PropTypes.func.isRequired,
  className: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired
}