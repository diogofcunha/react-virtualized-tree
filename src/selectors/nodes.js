import { createSelector } from 'reselect';
import { omit } from 'lodash';

import { UPDATE_TYPE } from '../contants';

export const getNodeRenderOptions = createSelector(
  (node => (node.state || {}).expanded),
  (node => (node.state || {}).favorite),
  (node => (node.state || {}).deletable),
  (node => node.children), 
  (expanded, favorite, deletable, children = []) => ({
    hasChildren: !!children.length,
    isExpanded: !!expanded,
    isFavorite: !!favorite,
    isDeletable: !!deletable
  })
);

const FLATTEN_TREE_PROPERTIES = [ 'deepness', 'parents' ];
const NODE_OPERATION_PROPERTIES = [ 'index' ];

const NODE_OPERATION_TYPES = {
  CHANGE_NODE: 'CHANGE_NODE',
  DELETE_NODE: 'DELETE_NODE',
  ADD_NODE: 'ADD_NODE'
}

const NODE_CHANGE_OPERATIONS = {
  CHANGE_NODE: (nodes, updatedNode) => nodes.map(n => n.id === updatedNode.id ? omit(updatedNode, FLATTEN_TREE_PROPERTIES) : n),
  DELETE_NODE: (nodes, updatedNode) => nodes.filter(n => n.id !== updatedNode.id),
  ADD_NODE: (nodes, nodeToAdd) => [ ...nodes, omit(nodeToAdd, [ ...FLATTEN_TREE_PROPERTIES, ...NODE_OPERATION_PROPERTIES] ) ]
}

export const replaceNodeFromTree = (nodes, updatedNode, operation = NODE_OPERATION_TYPES.CHANGE_NODE) => {
  if (!NODE_CHANGE_OPERATIONS[operation]) {
    return nodes;
  }
  
  const { parents } = updatedNode;

  if (!parents.length) {
    return NODE_CHANGE_OPERATIONS[operation](nodes, updatedNode);
  }

  const parentIndex = nodes.findIndex(n => n.id === parents[0])

  const preSiblings = nodes.slice(0, parentIndex);
  const postSiblings = nodes.slice(parentIndex + 1);

  return [
    ...preSiblings,
    {
      ...nodes[parentIndex], 
      ...nodes[parentIndex].children || operation === NODE_OPERATION_TYPES.ADD_NODE ? 
        { children: replaceNodeFromTree(
            nodes[parentIndex].children || [],
            { ...updatedNode, parents: parents.slice(1) },
            operation
          )
        } : {}
    },
    ...postSiblings
  ]
}

export const deleteNodeFromTree = (nodes, deletedNode) => {
  return replaceNodeFromTree(
    nodes,
    deletedNode,
    NODE_OPERATION_TYPES.DELETE_NODE
  );
}

export const addNodeToTree = (nodes, nodeToAdd) => replaceNodeFromTree(
  nodes,
  nodeToAdd,
  NODE_OPERATION_TYPES.ADD_NODE
)

export const moveNodeToChildren = (nodes, [ fromNode, toNode ]) => {
  const { id: targetId, parents: targetParents } = toNode;
  const parents = [ ...targetParents, ...targetId !== fromNode.id ? [targetId] : []];

  return moveNodeFromTree(nodes, [ fromNode, { ...toNode, parents } ]);
}

export const moveNodeFromTree = (nodes, [ fromNode, toNode ]) => {
  const treeWithoutMovedNode = deleteNodeFromTree(
    nodes,
    fromNode
  );

  return addNodeToTree(treeWithoutMovedNode, { ...fromNode, parents: toNode.parents });
}

export const udpateNode = (originalNode, newState) => ({
  node: {
    ...originalNode,
    state : {
      ...originalNode.state,
      ...newState
    }
  },
  type: UPDATE_TYPE.UPDATE
});

export const deleteNode = node => ({
  node,
  type: UPDATE_TYPE.DELETE
});

export const addNode = node => ({
  node,
  type: UPDATE_TYPE.ADD
});
