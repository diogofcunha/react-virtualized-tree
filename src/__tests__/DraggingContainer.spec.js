import React from 'react';
import { shallow } from 'enzyme';

import DraggingContainer from '../DraggingContainer';
import { Nodes } from '../../testData/sampleTree';
import DefaultGroupRenderer from '../filtering/DefaultGroupRenderer';
import { getFlattenedTree } from '../selectors/getFlattenedTree';
import { UPDATE_TYPE } from '../contants';

describe('DraggingContainer', () => {
  const getSampleNode = (i) => getFlattenedTree(Nodes)[i];

  const setup = (extraProps = {}, state) => {
    const props = { onChange: jest.fn(), className: '' };

    const wrapper = shallow(
      <DraggingContainer {...props}>
        <span></span>
      </DraggingContainer>
    );

    if (state) {
      wrapper.setState(state);
    }

    return {
      wrapper,
      props,
      createdContext: wrapper.instance().getChildContext()
    }
  }

  describe('Context', () => {
    it('should add to state on created handleDrag', () => {
      const fromNode = getSampleNode(2);

      const { createdContext, wrapper } = setup();

      createdContext.handleDrag(fromNode);

      expect(wrapper.state().fromNode).toEqual(fromNode);
    });

    it('should clear the state and call onChange with the correct params on created handleDrop', () => {
      const fromNode = getSampleNode(2);

      const { createdContext, wrapper, props } = setup();

      wrapper.setState({ fromNode });
      
      const toNode = getSampleNode(5);
      createdContext.handleDrop(toNode);

      expect(wrapper.state().fromNode).toEqual({});

      expect(props.onChange).toHaveBeenCalledWith({
        node: [ fromNode, toNode ],
        type: UPDATE_TYPE.MOVE
      });
    });

    it('should add node from state to context', () => {
      const { createdContext } = setup();

      expect(createdContext.fromNode).toEqual({});

      const fromNode = getSampleNode(2);
      const { createdContext: createdContextWithFromNode } = setup({}, { fromNode });

      expect(createdContextWithFromNode.fromNode).toEqual(fromNode);
    });
  });

  describe('Container DnD', () => {
    it('when something is dropped on the root, should call onChange with fromNode, without parents', () => {
      const fromNode = getSampleNode(2);

      const { wrapper, props } = setup({}, { fromNode });

      wrapper.first().simulate('drop');

      expect(props.onChange.mock.calls[0]).toMatchSnapshot();
    });

    it('drag over should prevent default', () => {
      const { wrapper } = setup();
      const event = { preventDefault: jest.fn() };

      wrapper.first().simulate('dragOver', event);

      expect(event.preventDefault).toHaveBeenCalled();
    });
  });
});
