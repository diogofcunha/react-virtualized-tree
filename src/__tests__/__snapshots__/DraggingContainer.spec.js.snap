// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggingContainer Container DnD when something is dropped on the root, should call onChange with fromNode, without parents 1`] = `
Array [
  Object {
    "node": Array [
      Object {
        "deepness": 2,
        "id": 3,
        "name": "Leaf 3",
        "parents": Array [
          0,
          2,
        ],
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
      Object {
        "deepness": 2,
        "id": 3,
        "name": "Leaf 3",
        "parents": Array [],
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
    ],
    "type": 4,
  },
]
`;
